File: test/Authzi.Tests.MicrosoftOrleans/MicrosoftEntra/MicrosoftEntraID.Common/Authorization/RoleBasedAuthorizationTests.fs
namespace Authzi.MicrosoftOrleans.AzureActiveDirectory.Tests.Authorization

open Authzi.MicrosoftOrleans.AzureActiveDirectory.Tests
open Authzi.MicrosoftOrleans.Grains.RoleBasedAuthorization
open Authzi.Security
open RootConfiguration
open System.Threading.Tasks
open Xunit

type RoleBasedAuthorizationTestsBase() =
    [<Theory>]
    [<MemberData(nameof(TestData.UserWithScopeAdeleV), MemberType = typeof<TestData>)>] 
    member _.``Multiple roles as a comma separated list should work when the user has both roles``
        (userName: string) (password: string) (scope: string list) =
        async {
            // Arrange
            let! accessToken = AccessTokenProvider.getAccessTokenForUserOnWebClient1Async
                                           userName password |> Async.AwaitTask

            let clusterClient = getClusterClient accessToken
            let userGrain = clusterClient.GetGrain<IManagerGrain>(userName)
            let! value = userGrain.GetWithCommaSeparatedRoles("Secret") |> Async.AwaitTask

            Assert.True(value.Equals "Secret")
        }

    [<Theory>]
    [<MemberData(nameof(TestData.UserWithScopeAlexW), MemberType = typeof<TestData>)>] 
    member _.``Multiple roles as a comma separated list shouldn't work when the user has only one role``
        (userName: string) (password: string) (scope: string list) =
        async {
            // Arrange
            let! accessToken = AccessTokenProvider.getAccessTokenForUserOnWebClient1Async
                                           userName password |> Async.AwaitTask

            let clusterClient = getClusterClient accessToken
            let userGrain = clusterClient.GetGrain<IManagerGrain>(userName)

            // Act
            let action =
                async {
                    let! value = userGrain.GetWithCommaSeparatedRoles("Secret") |> Async.AwaitTask
                    return value } |> Async.StartAsTask :> Task

            Assert.ThrowsAsync<AuthorizationException>(fun () -> action) |> ignore
        }

    [<Theory>]
    [<MemberData(nameof(TestData.UserWithScopeAdeleV), MemberType = typeof<TestData>)>]
    member _.``Multiple roles applied as multiple attributes should work when the user has both roles``
        (userName: string) (password: string) (scope: string list) =
        async {
            // Arrange
            let! accessToken = AccessTokenProvider.getAccessTokenForUserOnWebClient1Async
                                           userName password |> Async.AwaitTask

            let clusterClient = getClusterClient accessToken
            let userGrain = clusterClient.GetGrain<IManagerGrain>(userName)
            let! value = userGrain.GetWithMultipleRoleAttributes("Secret") |> Async.AwaitTask

            Assert.True(value.Equals "Secret")
        }

    [<Theory>]
    [<MemberData(nameof(TestData.UserWithScopeAlexW), MemberType = typeof<TestData>)>]
    member _.``Multiple roles applied as multiple attributes shouldn't work when the user has only one role``
        (userName: string) (password: string) (scope: string list) =
        async {
            // Arrange
            let! accessToken = AccessTokenProvider.getAccessTokenForUserOnWebClient1Async
                                           userName password |> Async.AwaitTask

            let clusterClient = getClusterClient accessToken
            let userGrain = clusterClient.GetGrain<IManagerGrain>(userName)

            // Act
            let action =
                async {
                    let! value = userGrain.GetWithMultipleRoleAttributes("Secret") |> Async.AwaitTask
                    return value } |> Async.StartAsTask :> Task

            Assert.ThrowsAsync<AuthorizationException>(fun () -> action) |> ignore
        }

File: test/AuthZI.Tests.MicrosoftOrleans/MicrosoftEntra/MicrosoftEntraID.Common.CSharp/Authorization/RoleBasedAuthorizationTests.cs
using System.Collections.Generic;
using System.Threading.Tasks;
using AuthZI.Security;
using AuthZI.Tests.MicrosoftOrleans.Grains.RoleBasedAuthorization;
using Xunit;

namespace AuthZI.Tests.MicrosoftOrleans.MicrosoftEntra.MicrosoftEntraID.Common.Authorization;

public class RoleBasedAuthorizationTestsBase
{
  [Theory]
  [MemberData(nameof(TestData.UserWithScopeAdeleV), MemberType = typeof(TestData), DisableDiscoveryEnumeration = true)]
  public async Task MultipleRolesAsACommaSeparatedListShouldWorkWhenTheUserHasBothRoles(
    string userName,
    string password,
    IEnumerable<string> scope)
  {
    var accessToken = await AccessTokenProvider.getAccessTokenForUserOnWebClient1Async(userName, password);

    var clusterClient = RootConfiguration.getClusterClient(accessToken);
    var userGrain = clusterClient.GetGrain<IManagerGrain>(userName);
    var value = await userGrain.GetWithCommaSeparatedRoles("Secret");

    Assert.True(value.Equals("Secret"));
  }

  [Theory]
  [MemberData(nameof(TestData.UserWithScopeAlexW), MemberType = typeof(TestData), DisableDiscoveryEnumeration = true)]
  public async Task MultipleRolesAsACommaSeparatedListShouldNotWorkWhenTheUserHasOnlyOneRole(
    string userName,
    string password,
    IEnumerable<string> scope)
  {
    var accessToken = await AccessTokenProvider.getAccessTokenForUserOnWebClient1Async(userName, password);

    var clusterClient = RootConfiguration.getClusterClient(accessToken);
    var userGrain = clusterClient.GetGrain<IManagerGrain>(userName);

    await Assert.ThrowsAsync<AuthorizationException>(() => userGrain.GetWithCommaSeparatedRoles("Secret"));
  }

  [Theory]
  [MemberData(nameof(TestData.UserWithScopeAdeleV), MemberType = typeof(TestData), DisableDiscoveryEnumeration = true)]
  public async Task MultipleRolesAppliedAsMultipleAttributesShouldWorkWhenTheUserHasBothRoles(
    string userName,
    string password,
    IEnumerable<string> scope)
  {
    var accessToken = await AccessTokenProvider.getAccessTokenForUserOnWebClient1Async(userName, password);

    var clusterClient = RootConfiguration.getClusterClient(accessToken);
    var userGrain = clusterClient.GetGrain<IManagerGrain>(userName);
    var value = await userGrain.GetWithMultipleRoleAttributes("Secret");

    Assert.True(value.Equals("Secret"));
  }

  [Theory]
  [MemberData(nameof(TestData.UserWithScopeAlexW), MemberType = typeof(TestData), DisableDiscoveryEnumeration = true)]
  public async Task MultipleRolesAppliedAsMultipleAttributesShouldNotWorkWhenTheUserHasOnlyOneRole(
    string userName,
    string password,
    IEnumerable<string> scope)
  {
    var accessToken = await AccessTokenProvider.getAccessTokenForUserOnWebClient1Async(userName, password);

    var clusterClient = RootConfiguration.getClusterClient(accessToken);
    var userGrain = clusterClient.GetGrain<IManagerGrain>(userName);

    await Assert.ThrowsAsync<AuthorizationException>(() => userGrain.GetWithMultipleRoleAttributes("Secret"));
  }
}
